app/src/main/java/com/lagradost/cloudstream3/ui/settings/SettingsAccount.kt

```diff
@@ -133,7 +142,85 @@ class SettingsAccount : PreferenceFragmentCompat(), BiometricAuthenticator.Biome
+                                        activity.runOnUiThread {
+                                            showToast(
+                                                activity.getString(R.string.device_pin_error_message)
+                                            )
+                                        }
+                                        api.authenticate(activity)
+                                        return@ioSafe
+                                    }
+
+                                    val qrCodeImage = QRCode.ofRoundedSquares()
+                                        .withColor(activity.colorFromAttribute(R.attr.colorPrimary))
+                                        .build(pinCodeData.verificationUrl)
+                                        .render().nativeImage() as Bitmap
+
+                                    activity.runOnUiThread {
+                                        dialog.show()
+                                        binding.devicePinCode.setText(txt(pinCodeData.userCode))
+                                        binding.deviceAuthMessage.setText(txt(R.string.device_pin_url_message, pinCodeData.verificationUrl))
+                                        binding.deviceAuthQrcode.setImage(
+                                            img(qrCodeImage)
+                                        )
+                                    }
+
+                                    var expirationCounter = pinCodeData.expiresIn
+
+                                    while (expirationCounter > 0) {
+                                        activity.runOnUiThread {
+                                            binding.deviceAuthValidationCounter.setText(
+                                                txt(
+                                                    R.string.device_pin_counter_text, expirationCounter.div(60) , expirationCounter.rem(60)
+                                                )
+                                            )
+                                        }
+
+                                        if (expirationCounter.rem(pinCodeData.interval)  == 0 && api.handleDeviceAuth(pinCodeData)) {
+                                            activity.runOnUiThread {
+                                                    showToast(
+                                                        activity.getString(R.string.authenticated_user)
+                                                            .format(
+                                                                api.name
+                                                            )
+                                                    )
+                                                dialog.dismissSafe(activity)
+                                            }
+                                            return@ioSafe
+                                        }
+                                        delay(1000)
```

Comment:
this entire setup is weird and prone to errors. Just use something like this https://developer.android.com/reference/android/os/CountDownTimer instead